whisper/audio.py

```diff
@@ -53,105 +51,175 @@ def load_audio(file: str, sr: int = SAMPLE_RATE):
-        if isinstance(audio, str):
-            audio = load_audio(audio)
-        audio = torch.from_numpy(audio)
-
-    if device is not None:
-        audio = audio.to(device)
-    if padding > 0:
-        audio = F.pad(audio, (0, padding))
-    window = torch.hann_window(N_FFT).to(audio.device)
-    stft = torch.stft(audio, N_FFT, HOP_LENGTH, window=window, return_complex=True)
-    magnitudes = stft[..., :-1].abs() ** 2
-
-    filters = mel_filters(audio.device, n_mels)
-    mel_spec = filters @ magnitudes
-
-    log_spec = torch.clamp(mel_spec, min=1e-10).log10()
-    log_spec = torch.maximum(log_spec, log_spec.max() - 8.0)
-    log_spec = (log_spec + 4.0) / 4.0
-    return log_spec
+    
+    # Cache for Hann windows based on size and device.
+    if not hasattr(get_hann_window, 'cache'):
+        get_hann_window.cache = {}
+    
+    key = (size, str(device))
+    if key not in get_hann_window.cache:
+        get_hann_window.cache[key] = torch.hann_window(size).to(device)
+    
+    return get_hann_window.cache[key]
+
+def log_mel_spectrogram(
+   audio: Union[str, np.ndarray, torch.Tensor],
+   n_mels: int = 80,
+   padding: int = 0,
+   device: Optional[Union[str, torch.device]] = None,
+) -> torch.Tensor:
+   """
+   Compute the log-Mel spectrogram of an audio waveform.
+
+   Parameters
+   ----------
+   audio: Union[str, np.ndarray, torch.Tensor]
+       The path to audio or a NumPy array or Tensor containing the audio waveform.
+
+   n_mels: int
+       The number of Mel-frequency filters (only supports 80 and 128).
+
+   padding: int
+       Number of zero samples to pad to the right.
+
+   device: Optional[Union[str, torch.device]]
+       If given, moves the audio tensor to this device before STFT.
+
+   Returns
+   -------
+   torch.Tensor
+       A Tensor containing the log-Mel spectrogram.
+   """
+   
+   # Load audio if necessary and convert to tensor if needed.
+   if isinstance(audio, str):
+       audio_tensor = load_audio(audio)
+       audio_tensor = torch.from_numpy(audio_tensor)
+   else:
+       audio_tensor = audio if isinstance(audio, torch.Tensor) else torch.from_numpy(audio)
+
+   # Move to specified device if provided.
+   if device is not None:
+       audio_tensor = audio_tensor.to(device)
+
+   # Padding the audio tensor.
+   if padding > 0:
+       audio_tensor = F.pad(audio_tensor, (0, padding))
+
+   # Compute STFT and magnitudes.
+   window = get_hann_window(AudioConstants.N_FFT, audio_tensor.device)
+   stft = torch.stft(audio_tensor, AudioConstants.N_FFT,
+                      AudioConstants.HOP_LENGTH,
+                      window=window,
+                      return_complex=True)
+   
+   magnitudes = stft.abs() ** 2
+
+   # Calculate Mel spectrogram and apply logarithmic scaling.
+   filters = mel_filters(audio_tensor.device, n_mels)
+   mel_spec = filters @ magnitudes
+
+   log_spec = torch.clamp(mel_spec, min=1e-10).log10()
+   
+   log_spec_normalized = (log_spec + 4.0) / 4.0
```

Comment:
maybe combine both?
log_spec_normalized = (torch.clamp(mel_spec, min=1e-10).log10() + 4.0) / 4.0

Reply:
@Nisarg236 I did it separately to try to keep the code clearer, but yes, it is possible to combine the two.